src/storage-controller/src/lib.rs

```diff
@@ -797,17 +797,59 @@ where
+                    let metadata = self.storage_collections.collection_metadata(id)?.clone();
+                    let persist_client = self
+                        .persist
+                        .open(metadata.persist_location.clone())
+                        .await
+                        .unwrap();
+
+                    let read_handle_fn = move || {
+                        let persist_client = persist_client.clone();
+                        let metadata = metadata.clone();
+
+                        let fut = async move {
+                            let read_handle = persist_client
+                                .open_leased_reader::<SourceData, (), T, Diff>(
+                                    metadata.data_shard,
+                                    Arc::new(metadata.relation_desc.clone()),
+                                    Arc::new(UnitSchema),
+                                    Diagnostics {
+                                        shard_name: id.to_string(),
+                                        handle_purpose: format!("snapshot {}", id),
+                                    },
+                                    USE_CRITICAL_SINCE_SNAPSHOT.get(persist_client.dyncfgs()),
+                                )
+                                .await
+                                .expect("invalid persist usage");
+                            read_handle
+                        };
+
+                        fut.boxed()
+                    };
+
+                    // How to decide between collection types:
+                    //
+                    // Collection at any given time is derived from some state
+                    // at that time. In a way, it's a view on top of something
+                    // -> incremental collection
+                    //
+                    // Collection records things that happen over time, like a
+                    // history of something -> append-only collection.
 
                     match i {
                         IntrospectionType::ShardMapping => {
+                            self.collection_manager.register_incremental_collection(id, read_handle_fn);
                             self.initialize_shard_mapping().await;
                         }
                         IntrospectionType::Frontiers | IntrospectionType::ReplicaFrontiers => {
-                            // Set the collection to empty.
-                            self.reconcile_managed_collection(id, vec![]).await;
+                            self.collection_manager.register_incremental_collection(id, read_handle_fn);
+                            // Incremental collections start with an empty
+                            // desired state. No need to manually reset.
                         }
                         IntrospectionType::StorageSourceStatistics => {
+                            self.collection_manager.register_incremental_collection(id, read_handle_fn);
```

Comment:
I'm not entirely sure that the stats should be modeled as incremental collections. I kind of think you might just want to toss the stats until the read-write promotion. Otherwise the cutover from the old version to the new version might cause the stats to regress, which will be quite surprising.